File: Algorithms/Sorting/QuickSort.java
import java.util.Arrays;

public class QuickSort {
    public static void main(String[] args) {
        int[] arr = { 5, 1, 3, 2, 4 };
        System.out.println("Unsorted: " + Arrays.toString(arr));
        quickSort(arr, 0, arr.length - 1);
        System.out.println("Sorted: " + Arrays.toString(arr));
    }

    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // {5,1,3,2,4}
            // {1,3,2,4,5}
            int pi = partition(arr, low, high);
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1; // pivot index in the sorted list

        // {5,1,3,2,4}
        // {1,3,2,5,4}
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                int temp = arr[j];
                arr[j] = arr[i];
                arr[i] = temp;
            }
        }

        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }
}

File: Algorithms/Sorting/QuickSort.js
const arr = [5, 4, 3, 2, 1, 4];

// swaps two elements in an arr using their index
const swap = (arr, x, y) => {
  let temp = arr[x];
  arr[x] = arr[y];
  arr[y] = temp;
};

// partion around the pivot element.
const partition = (arr, left, right) => {
  let pivot = arr[right];
  let pivotIndex = left - 1;

  for (let i = left; i < right; i++) {
    if (arr[i] < pivot) {
      pivotIndex++;
      swap(arr, i, pivotIndex);
    }
  }

  pivotIndex++;
  swap(arr, pivotIndex, right);
  return pivotIndex;
};

// apply quickSort on both halves
const quickSort = (arr, left, right) => {
  if (left < right) {
    let pi = partition(arr, left, right);

    quickSort(arr, left, pi - 1);
    quickSort(arr, pi + 1, right);
  }
};

console.log("Arr: ", arr);
quickSort(arr, 0, arr.length - 1);
console.log("Quick sorted:", arr);
